### DISCUSSION TITLE
Soroban vs NEAR: non-trivial contract sizes & DevX

### DISCUSSION THREAD

**Discussion by chadoh** [stellar/sorobanathon#26] [General]
[Comparing Hello World examples](https://github.com/stellar/sorobanathon/discussions/19) is fine, but as multiple commenters pointed out, how do NEAR and Soroban stack up with more complex examples?

TL;DR: Soroban contracts are still smaller, but not 100x like a simple Hello World. More like 4x, reducible to ~2x by motivated NEAR/Rust experts. And that quadrupled size on NEAR? Might be worth it!

# Fully unoptimized: NEAR wins by 4x!

The `near-sdk-rs` source code contains [examples](https://github.com/near/near-sdk-rs/tree/master/examples), hard-wired to the latest greatest SDK features. If you clone the SDK source and `cd` into the `examples/fungible-token` directory and build:

    cargo build --target wasm32-unknown-unknown --release
    
The resulting Wasm: 178K

![](https://i.imgur.com/TjVDoWL.png)

If you clone the [Soroban token example](https://github.com/stellar/soroban-token-contract) and build it the same way:

744K!

![](https://i.imgur.com/LdoeanY.png)

# OK, but with simple compilation optimizations?

Without changing any code, how small do these get with a smarter build?

For NEAR, we can build with [RAEN](https://raen.dev/):

```bash 
cargo install raen
raen build --release
```

RAEN adds extra info (the contract's schema/interface/abi) to the .wasm file, but still gives you a smaller file size (thanks, optimizations and compression!). This gets the contract to:

110K:

![](https://i.imgur.com/AA8dVpz.png)

For Soroban, we can build with [the recommended optimizations](https://soroban.stellar.org/docs/tutorials/build-optimized):

```bash 
cargo +nightly build \
    --target wasm32-unknown-unknown \
    --release \
    -Z build-std=std,panic_abort \
    -Z build-std-features=panic_immediate_abort
    
wasm-opt -Oz \
    target/wasm32-unknown-unknown/release/soroban_token_contract.wasm \
    -o target/wasm32-unknown-unknown/release/soroban_token_contract_optimized.wasm
```

This gets the contract down to 27K!

![](https://i.imgur.com/8zaGgg9.png)

# Aside: 744K to 27K???

Ridiculous.

# Why are NEAR contracts bigger?

JSON. [`serde_json`](https://docs.rs/serde_json), specifically.

When you call a NEAR contract, you provide the arguments with JSON:

    near call some-token.near ft_transfer '{
      "receiver_id": "you.near",
      "amount": "1000000000000000000000000"
    }' --account-id me.near

It's nice.

(Notice that amounts are also provided as strings, which means the string-parsing bits of the standard library are also essential to a NEAR smart contract.)

This nice developer experience comes at the cost of needing every smart contract to include JSON-parsing logic. By default, `near-sdk-rs` handles this JSON parsing with `serde_json`. It adds some bloat, but stays out of the smart contract developer's way, so you can focus on contract logic, not parsing inputs.

# You *can* make NEAR contracts without JSON

Or with less JSON bloat: you, the smart contract developer, could swap `serde_json` for [microjson](https://docs.rs/microjson). Then you *do* need to worry about parsing inputs; see [an example](https://github.com/AhaLabs/contract-riffs/blob/decd720642222b4c14c9e342f97b056eed980ee7/src/lib.rs#L71-L80).

This might be worth it for certain contracts & projects, but will probably never become a wide-spread practice.

You can even get rid of JSON arguments altogether. You can [serialize the inputs](https://github.com/aurora-is-near/rainbow-bridge/blob/bc165982786a058702d05016b6a3bb093aff75ca/contracts/near/eth-prover/src/lib.rs#L38-L41) with [Borsh](https://borsh.io/), or even parse the inputs as a base64-encoded string. But this makes it harder for people to interact with your contract. And you can't implement any [contract standards](https://nomicon.io/Standards/) this way; they all use JSON arguments.

# Conclusion: When is smaller file size worth it?

For this non-trivial example, we saw that Soroban contracts are still smaller than NEAR contracts by a factor of about 4x, or 2x for NEAR contracts that have some significant space-saving mods.

4x is a big difference, but not the breathtaking 100x we saw with [the Hello World comparison](https://github.com/stellar/sorobanathon/discussions/19#discussioncomment-3935546).

In fact, it's close enough that at this point we have to ask if the smaller size is still worth it. 

* Calling NEAR contracts has that nice JSON interface.
* NEAR contracts can parse input strings longer than 10 characters. (Upsetting that Soroban can't, tbqh!)
* Just strings in general! You can use them when programming NEAR smart contracts. You [can't](https://soroban.stellar.org/docs/learn/environment-concepts#values-and-types), with Soroban.
* NEAR contract developers get a [memory allocator](https://soroban.stellar.org/docs/learn/rust-dialect#limited-ideally-zero-dynamic-memory-allocation) by default, rather than needing to learn new embedded systems programming patterns.

So right now, I'm thinking NEAR's trade-offs are worth it.

# Well, but maybe...

The biggest downside of NEAR's developer experience might be its [asynchronous block execution](https://docs.near.org/develop/contracts/crosscontract), which burdens contract authors with unwinding failed transactions. This creates enormous surface area for bugs and security exploits.

Soroban has synchronous cross-contract calls like Ethereum, making this whole category of bugs go away.

But NEAR's asynchronous blocks are there for a reason. They're a key building block of NEAR's [sharding architecture](https://near.org/blog/near-launches-simple-nightshade-the-first-step-towards-a-sharded-blockchain/), which is how NEAR will stay fast and available as demand grows, and part of how NEAR incentivizes a large, decentralized group of hardware providers—you don't need to run the full blockchain in order to participate.

Soroban has its nice preflight + threading design (see the "Storage Footprint and Preflight" section in the ["Advanced" box here](https://soroban.stellar.org/docs/learn/interacting-with-contracts#example), so maybe sharding will not be needed nearly as soon—you can have higher throughput with those smaller, faster contracts! And since Stellar [cares less](https://developers.stellar.org/docs/run-core-node/tier-1-orgs) about full hardware-level decentralization, maybe throughput can eventually be increased further by requiring more beefy machines, a la Solana.

So *if* Soroban's trade-offs are there to allow it to scale _without_ sharding (is this even possible?? I asked about sharding/scaling plans in [Discord](https://discord.gg/6ZCcYqbC96) but didn't get a response), then I think the trade-offs are worth it. Better to have synchronous cross-contract calls, even if it requires learning new embedded-programming-style paradigms.

**Comment by chadoh** [comment]
@rice2000 I'm told you might be a good person to ask about a claim I make near the end of this post:

> ...since Stellar [cares less](https://developers.stellar.org/docs/run-core-node/tier-1-orgs) [than NEAR] about full hardware-level decentralization, maybe throughput can eventually be increased further by requiring more beefy machines, a la Solana.

First of all, do you think this is a good and fair way to represent things?

Second of all, do you think this, https://developers.stellar.org/docs/run-core-node/tier-1-orgs, is the best link to help people learn more about Stellar's approach to decentralization?

@tyvdh, who I DM'd on Discord, thinks maybe one of these would be better:

- https://stellar.org/papers/stellar-consensus-protocol
- https://www.stellar.org/developers-blog/intuitive-stellar-consensus-protocol?locale=en
- https://medium.com/interstellar/understanding-the-stellar-consensus-protocol-423409aad32e
- https://developers.stellar.org/docs/fundamentals-and-concepts/stellar-consensus-protocol

**Reply by rice2000** [reply]
Good question! 

Generally, Stellar's design favors giving network operators control over network operating costs by allowing them to vote on network configurations.  

For instance, validators vote on the maximum number of txs / ledger.  Currently, it's set to 1,000 ops/ledger, lower than any theoretical limits.  Why keep headroom?  Validators chose the current limit to allow enough throughput to support Stellar use cases, but to keep hardware requirements low enough to allow people all over the world to keep up (including people who run Horizon nodes to submit transactions and query network data).  It's an attempt to balance performance with inclusiveness. 

In general, a good place to look for Stellar network goals and development values is the README for the Core Advancement Proposal repo.: https://github.com/stellar/stellar-protocol/tree/master/core#stellar-network-goals

In terms of a specific plan to deal with ledger growth, there is an open discussion around a specific CAP designed to create a fee model for smart contracts: https://github.com/stellar/stellar-protocol/blob/master/core/cap-0055.md

If you scroll down, you'll find a section called "Ledger size reduction": https://github.com/stellar/stellar-protocol/blob/master/core/cap-0055.md#ledger-size-reduction

It's all still a WIP, and still actively being discussed on the Stellar Dev mailing list: https://groups.google.com/g/stellar-dev/c/0mQ5E8GXK_M/m/gg6dV2qPAQAJ

The resources Tyler suggested are good for understanding how the Stellar Consensus Protocol works, although I prefer this updated paper to the original SCP paper.  I think it's a better read: https://www.stellar.org/papers/fast-and-secure-global-payments-with-stellar

The Tier 1 section you link to is really just a guide to best practices.  SCP relies on voting, and validator operators choose which other validators to add to their quorum sets, and over the years, a set of common expectations started to emerge, and so we tried to capture it in the documentation.

Does that answer your question?

**Comment by leighmcculloch** [comment]
This is very interesting. The token example contract heavily uses string errors, which is why its size is so much larger when not built with the nightly build that strips them all out. We recently a way to have non-string errors, and I expect if we revised that example to use them the size would be much lower from the get go.

**Comment by kalepail** [comment]
> Just strings in general! You can use them when programming NEAR smart contracts. 
> You [can't](https://soroban.stellar.org/docs/learn/environment-concepts#values-and-types), with Soroban.

I found this to be quite challenging to work around myself. I wonder if we should add more native `String` support into Soroban as a convenience or if this begins to chip away at our advantages over alternatives? Efficiency and safety over convenience. 🤔 

Either way great write up! 👏

**Reply by leighmcculloch** [reply]
@tyvdh I'd like to learn more about the use cases here. Strings are for the most part bytes, and we can definitely store and use bytes in Soroban contracts. What operations are missing? Is it a presentation issue?

Even if we add strings they will probably appear as bytes in contracts and the "string" quality will be mostly symbolic for rendering.

**Reply by kalepail** [reply]
Definitely more as a utility in the same way the std library does it. (Or I assume it does it)

I’m not convinced it’s worth as byte conversion is quite easy and understood. Would be curious to hear @chadoh thoughts on this. It’s the fine balance of convenience at the expense of contract size. If one of our selling points is compact by default we need to be careful backing out of that. Though having opt in utilities may be nice. So you could use bytes if you wanted but a String helper type exists if you want to import it. I guess the question would be would just the fact of having the type be an option require all contract sizes be just that slight bit larger.

**Reply by leighmcculloch** [reply]
Having our own string type wouldn't increase contract size by much. The main reason we don't have the Rust `String` is it requires an allocator. Separate to the representation of things outside of the contract, in a contract what things do you see needing strings for that a byte slice can't do?

@graydon Thoughts on adding a `String` type that is essentially same interface as the `Bytes` interface? The distinction the type would introduce would make it easier for tools and off-chain systems to know that the value is intended to be a string and could render it as so.

**Reply by graydon** [reply]
I remain 100% in favour of having a distinct String type (whether or not we insist it's UTF-8-only; I would do that but I understand opinions vary). I added String as a type initially but it was something early feedback (eg. https://groups.google.com/g/stellar-dev/c/X0oRzJoIr10) suggested was unwanted / over-engineering. I remain interested in re-adding them, in some form.

(I also think we might be able to have a first-class _slice_ type that can point into either a `Bytes`, `String` or guest memory, without having to allocate fresh objects. It'd need either a wider RawVal type or to place a few fairly harmless size restrictions on linear memory and total host object count)

**Comment by achallagundla** [comment]
Heya Chad! 
Thanks for the submission, will get back to you soon after review!

**Comment by achallagundla** [comment]
Heya Chad,
Congratulations! Your contribution to Sorobanathon: First Light has been accepted! 🎉
Your submission may qualify for a reward of **1200 XLM**! One last step: I just need you to confirm your public key is still the same as what you previously submitted, will message you on discord. 

Please let us know if you have any questions. (You can also message me on discord @ anuxhya)

All the best,
Anuhya

**Comment by leighmcculloch** [comment]
Just realized we accidentally are disabling LTO optimizations on our builds because of a crate-type config we had set unnecessarily in some of our builds. Building the token example contract here now comes out as 19K: https://github.com/stellar/soroban-examples/pull/164
